File: tools/webcheck.mjs
#!/usr/bin/env node
// Web smoke check for the libwgrender examples (tools/verify.py --web runs it).
//
// Serves build/web-<backend> with tools/serve.py, loads each built example in a
// Chromium-based browser (Brave, Chrome, Chromium) through the DevTools
// protocol, and fails an example if it logs a console error or a libwgrender
// [ERROR]/[FATAL] line, throws, hits a sokol panic, never reports starting on the
// expected backend, or is still loading assets when its time runs out. A screenshot
// of every example is saved for a visual check.
//
// No npm dependencies: needs Node >= 22 (built-in fetch and WebSocket).
//
//   node tools/webcheck.mjs [options] [example ...]     (default: all built examples)
//
//   --backend=webgl2|webgpu  backend to check (default webgl2); the site is
//                       build/web-<backend> (the CMake preset of that name)
//   --headed            show the browser window on the real screen. Otherwise WebGL2
//                       runs headless, and WebGPU (which gets no working GPU device
//                       headless) runs on a private virtual X display (Xvfb, ANGLE on
//                       Vulkan), so it never shows a window or wakes the monitors. Without
//                       Xvfb installed, WebGPU falls back to the real screen.
//   --settle=MS         longest an example runs before it is checked (default 20000). An
//                       example is checked once it has started, has no asset tasks
//                       pending (libwgrender's queue) and no network requests in flight, and
//                       has run --quiet ms since the last of those changed
//   --quiet=MS          (default 1500)
//   --jobs=N            examples checked at once (default 4). Each example has its own
//                       browser context (own storage; its own window when headed)
//   --out=DIR           screenshot directory (default build/web-<backend>/webcheck)
//   --browser=PATH      browser executable (or WEBCHECK_BROWSER; default: first
//                       found of brave-browser-stable, google-chrome-stable,
//                       google-chrome, chromium, chromium-browser)
//   --verbose           print every console line and browser log entry an example
//                       produced (WebGPU validation messages arrive as log entries,
//                       not console calls, so this is how to see them)
//
// Cleanup: the browser and server are always stopped, including when this script
// crashes or is killed (see RunProcesses in weblib.mjs).
//
// Note: never call canvas.getContext() from here. A canvas that already has a
// WebGL context can't be used for WebGPU, which breaks the example under test.
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { basename, join } from "node:path";

import { findBrowser, findXvfb, freePort, launchBrowser, openSession, ROOT, RunProcesses, sleep, waitFor } from "./weblib.mjs";

const BACKEND_LOG = { webgl2: "GLES3/WebGL2 backend", webgpu: "WebGPU backend" };

function parseArgs(argv) {
    const opts = { backend: "webgl2", headed: false, settle: 20000, quiet: 1500, jobs: 0, out: null,
                   browser: process.env.WEBCHECK_BROWSER, threads: true, verbose: false, examples: [] };
    for (const arg of argv) {
        const [key, value] = arg.split(/=(.*)/s);
        switch (key) {
            case "--backend": opts.backend = value; break;
            case "--headed": opts.headed = true; break;
            case "--settle": opts.settle = Number(value); break;
            case "--quiet": opts.quiet = Number(value); break;
            case "--jobs": opts.jobs = Number(value); break;
            case "--out": opts.out = value; break;
            case "--browser": opts.browser = value; break;
            case "--threads": opts.threads = value !== "0"; break;
            case "--verbose": opts.verbose = true; break;
            default:
                if (arg.startsWith("--")) throw new Error(`unknown option ${arg}`);
                opts.examples.push(arg);
        }
    }
    if (!(opts.backend in BACKEND_LOG)) throw new Error(`--backend must be webgl2 or webgpu, got '${opts.backend}'`);
    /* where the browser shows its windows: "headless", "xvfb" or "screen" */
    opts.display = opts.headed ? "screen" : opts.backend === "webgpu" ? (findXvfb() ? "xvfb" : "screen") : "headless";
    opts.site = join(ROOT, "build", `web-${opts.backend}${opts.threads ? "" : "-nothreads"}`);
    if (!(opts.jobs >= 1)) opts.jobs = 4;
    opts.out ??= join(opts.site, "webcheck");
    return opts;
}

async function checkExample(browser, debugBase, baseUrl, example, opts) {
    const result = { example, errors: [], started: false, backendOk: false, screenshot: null, console: [] };
    /* Each example gets its own browser context (separate storage, like a fresh
     * profile), so examples checked at the same time don't share the IndexedDB
     * file cache and can't slow or affect each other. */
    const { browserContextId } = await browser.send("Target.createBrowserContext", { disposeOnDetach: true });
    const { targetId } = await browser.send("Target.createTarget", { url: "about:blank", browserContextId });
    const session = await openSession(`ws://${new URL(debugBase).host}/devtools/page/${targetId}`);
    try {
        const inflight = new Set();
        let navigated = Date.now();
        let lastActivity = Date.now();
        session.onEvent((msg) => {
            if (msg.method === "Network.requestWillBeSent") {
                /* a worker's own script (threaded builds start pthread workers) has no
                 * loader, and its completion is reported to the worker, not this page */
                if (!msg.params.loaderId) return;
                inflight.add(msg.params.requestId);
                lastActivity = Date.now();
            } else if (msg.method === "Network.loadingFinished" || msg.method === "Network.loadingFailed") {
                inflight.delete(msg.params.requestId);
                lastActivity = Date.now();
            } else if (msg.method === "Runtime.consoleAPICalled") {
                const text = msg.params.args.map((a) => a.value ?? a.description ?? "").join(" ");
                result.console.push(text.trim().split("\n")[0]);
                if (text.includes("libwgrender:") && text.includes("backend")) {
                    result.started = true;
                    result.startMs ??= Date.now() - navigated;
                    lastActivity = Date.now();
                    result.backendOk ||= text.includes(BACKEND_LOG[opts.backend]);
                }
                /* libwgrender logs go to the console as plain messages: fail on error-level
                 * ones like tools/smoke.sh does ([ERROR], [FATAL]) */
                if (msg.params.type === "error" || text.includes("[panic]") || /\[(ERROR|FATAL)/.test(text)) {
                    result.errors.push(text.trim().split("\n")[0]);
                }
            } else if (msg.method === "Runtime.exceptionThrown") {
                const d = msg.params.exceptionDetails;
                result.errors.push((d.exception?.description || d.text || "exception").split("\n")[0]);
            } else if (msg.method === "Log.entryAdded") {
                /* the browser's own messages: WebGPU validation errors from Dawn, GL
                 * driver warnings, network failures. Error-level ones fail the check,
                 * except network ones: a missing favicon is a 404 too, and a missing
                 * asset already fails through libwgrender's own loading errors */
                const e = msg.params.entry;
                const line = `[${e.source}/${e.level}] ${(e.text || "").trim().split("\n")[0]}`;
                result.console.push(line);
                if (e.level === "error" && e.source !== "network") result.errors.push(line);
            }
        });
        await session.send("Runtime.enable");
        await session.send("Log.enable");
        await session.send("Page.enable");
        await session.send("Network.enable");
        const deadline = Date.now() + opts.settle;
        navigated = Date.now();
        await session.send("Page.navigate", { url: `${baseUrl}/?ex=${encodeURIComponent(example)}` });
        /* Wait until the example has started, its asset downloads are done, and it has
         * run for a while since (errors from loading show up in that window), or until
         * the deadline, whichever comes first. */
        let pending = -1;
        while (Date.now() < deadline) {
            if (!result.started) {
                /* don't call into the page before libwgrender reports it's running: calling an
                 * exported function before the wasm runtime is initialized aborts the page */
                await sleep(100);
                continue;
            }
            const { result: value } = await session.send("Runtime.evaluate", {
                expression: "typeof Module !== 'undefined' && Module._wgri_asset_pending_count ? Module._wgri_asset_pending_count() : -1",
                returnByValue: true,
            });
            if (value.value !== pending) {
                pending = value.value;
                lastActivity = Date.now();
            }
            if (result.started && pending === 0 && inflight.size === 0 && Date.now() - lastActivity >= opts.quiet) {
                break;
            }
            await sleep(100);
        }
        result.pending = pending;
        if (pending > 0) { /* out of time: have libwgrender say which files, and where they are stuck */
            await session.send("Runtime.evaluate", {
                expression: "typeof Module !== 'undefined' && Module._wgri_asset_pending_log && Module._wgri_asset_pending_log()",
            }).catch(() => {});
            await sleep(300); /* the warnings arrive as console events */
        }
        const shot = await session.send("Page.captureScreenshot", { format: "png" });
        result.screenshot = join(opts.out, `${example}.png`);
        writeFileSync(result.screenshot, Buffer.from(shot.data, "base64"));
    } finally {
        session.close();
        await browser.send("Target.closeTarget", { targetId }).catch(() => {});
        await browser.send("Target.disposeBrowserContext", { browserContextId }).catch(() => {});
    }
    return result;
}

async function main() {
    if (typeof WebSocket === "undefined") {
        throw new Error(`Node ${process.version} has no built-in WebSocket; webcheck needs Node >= 22`);
    }
    const opts = parseArgs(process.argv.slice(2));
    const manifest = join(opts.site, "examples.json");
    if (!existsSync(manifest)) {
        throw new Error(`no web build at ${opts.site} (build it first: cmake --preset ${basename(opts.site)} && cmake --build --preset ${basename(opts.site)})`);
    }
    const built = JSON.parse(readFileSync(manifest, "utf8"));
    const examples = opts.examples.length ? opts.examples : built;
    const missing = examples.filter((e) => !built.includes(e));
    if (missing.length) throw new Error(`not built: ${missing.join(", ")}`);

    const browserPath = findBrowser(opts.browser);
    mkdirSync(opts.out, { recursive: true });
    const run = new RunProcesses("webcheck");
    process.on("exit", () => run.stopNow());
    for (const [signal, code] of [["SIGINT", 130], ["SIGTERM", 143], ["SIGHUP", 129]]) {
        process.on(signal, () => { run.stop().finally(() => process.exit(code)); });
    }
    process.on("uncaughtException", (err) => {
        console.error(`webcheck: ${err.stack || err.message}`);
        run.stopNow();
        process.exit(2);
    });

    try {
        const sitePort = await freePort();
        run.spawn("python3", [join(ROOT, "tools", "serve.py"), String(sitePort), opts.site]);
        const baseUrl = `http://127.0.0.1:${sitePort}`;
        await waitFor(`${baseUrl}/examples.json`, "tools/serve.py");

        if (opts.display === "screen" && opts.backend === "webgpu" && !opts.headed) {
            console.log("webcheck: Xvfb not found; WebGPU runs in a window on the real screen");
        }
        const { debugBase, browser } = await launchBrowser(run, browserPath, { display: opts.display, backend: opts.backend });

        console.log(`webcheck: ${examples.length} example(s), backend ${opts.backend}, ` +
                    `${opts.display === "xvfb" ? "virtual display (Xvfb)" : opts.display}, ${opts.jobs} at a time, ${browserPath}`);
        /* check `jobs` examples at a time; report in order as results complete */
        const results = new Array(examples.length);
        let next = 0, reported = 0, failed = 0;
        const report = () => {
            while (reported < examples.length && results[reported]) {
                const r = results[reported++];
                const problems = [...r.errors];
                if (!r.started) {
                    problems.push("never logged its backend (did not start?); last console output:");
                    for (const line of r.console.slice(-6)) problems.push(`  | ${line}`);
                }
                else if (r.pending < 0) problems.push(`never reported its asset queue in ${opts.settle} ms (the runtime was still starting)`);
                else if (r.pending !== 0) {
                    problems.push(`still loading after ${opts.settle} ms (${r.pending} asset task(s) pending):`);
                    const stuck = r.console.filter((line) => line.includes("wgr_asset: pending:"));
                    for (const line of (stuck.length ? stuck : r.console.slice(-6))) problems.push(`  | ${line}`);
                }
                else if (!r.backendOk) problems.push(`started on a different backend than '${opts.backend}' (stale build?)`);
                if (problems.length) failed++;
                const start = r.startMs !== undefined ? ` (started after ${(r.startMs / 1000).toFixed(1)} s)` : "";
                console.log(`  ${problems.length ? "FAIL" : "ok  "}  ${r.example}${problems.length ? start : ""}`);
                for (const p of problems) console.log(`          ${p}`);
                if (opts.verbose) for (const line of r.console) console.log(`          | ${line}`);
            }
        };
        const worker = async () => {
            while (next < examples.length) {
                const index = next++;
                try {
                    results[index] = await checkExample(browser, debugBase, baseUrl, examples[index], opts);
                } catch (err) {
                    results[index] = { example: examples[index], errors: [`check failed: ${err.message}`],
                                       started: true, backendOk: true, pending: 0 };
                }
                report();
            }
        };
        await Promise.all(Array.from({ length: Math.min(opts.jobs, examples.length) }, worker));
        browser.close();
        console.log(`screenshots: ${opts.out}`);
        console.log(failed ? `FAIL: ${failed} of ${examples.length} example(s)` : `PASS: ${examples.length} example(s)`);
        return failed ? 1 : 0;
    } finally {
        await run.stop();
    }
}

main().then((code) => process.exit(code), (err) => {
    console.error(`webcheck: ${err.message}`);
    process.exit(2);
});

File: tools/webcheck.py
#!/usr/bin/env python3
"""Web smoke check for the libwgrender examples (tools/verify.py --web runs it).

    tools/webcheck.py [options] [example ...]     (default: all built examples)

Serves build/web-<backend> with tools/serve.py, loads each built example in a
Chromium-based browser (Brave, Chrome, Chromium, Edge) through the DevTools protocol,
and fails an example if it logs a console error or a libwgrender [ERROR]/[FATAL] line,
throws, hits a sokol panic, never reports starting on the expected backend, or is
still loading assets when its time runs out. A screenshot of every example is saved
for a visual check. Standard library only (tools/weblib.py).

  --backend=webgl2|webgpu  backend to check (default webgl2); the site is
                      build/web-<backend> (the CMake preset of that name)
  --threads=0         the -nothreads build
  --headed            show the browser window on the real screen. Otherwise WebGL2 runs
                      headless, and WebGPU (which gets no working GPU device headless)
                      runs on a private virtual X display (Xvfb, ANGLE on Vulkan), so it
                      never shows a window or wakes the monitors. Without Xvfb (on
                      Windows and macOS always), WebGPU runs on the real screen.
  --settle=MS         longest an example runs before it is checked (default 20000). An
                      example is checked once it has started, has no asset tasks pending
                      (libwgrender's queue) and no network requests in flight, and has run
                      --quiet ms since the last of those changed
  --quiet=MS          (default 1500)
  --jobs=N            examples checked at once (default 4). Each example has its own
                      browser context (own storage; its own window when headed)
  --out=DIR           screenshot directory (default build/web-<backend>/webcheck)
  --browser=PATH      browser executable (or WEBCHECK_BROWSER; default: the first of
                      Brave, Chrome, Chromium and Edge found on PATH or where they install)
  --verbose           print every console line and browser log entry an example
                      produced (WebGPU validation messages arrive as log entries, not
                      console calls, so this is how to see them)

The browser and server are always stopped, including when this is interrupted or
killed (RunProcesses in tools/weblib.py).

Never call canvas.getContext() from here: a canvas that already has a WebGL context
can't be used for WebGPU, which breaks the example under test.
"""
import argparse
import base64
import json
import re
import sys
import threading
import time
import urllib.parse
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))  # an embedded Python (Windows) doesn't add it
from weblib import (PYTHON, ROOT, RunProcesses, find_browser, find_xvfb, free_port, launch_browser,
                    open_session, wait_for)

BACKEND_LOG = {'webgl2': 'GLES3/WebGL2 backend', 'webgpu': 'WebGPU backend'}
ERROR_LINE = re.compile(r'\[(ERROR|FATAL)')


def parse_args():
    ap = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument('--backend', default='webgl2', choices=sorted(BACKEND_LOG))
    ap.add_argument('--threads', default='1', choices=['0', '1'])
    ap.add_argument('--headed', action='store_true')
    ap.add_argument('--settle', type=int, default=20000)
    ap.add_argument('--quiet', type=int, default=1500)
    ap.add_argument('--jobs', type=int, default=4)
    ap.add_argument('--out')
    ap.add_argument('--browser')
    ap.add_argument('--verbose', action='store_true')
    ap.add_argument('examples', nargs='*')
    opts = ap.parse_args()
    # where the browser shows its windows: headless, xvfb or screen
    opts.display = ('screen' if opts.headed else
                    ('xvfb' if find_xvfb() else 'screen') if opts.backend == 'webgpu' else 'headless')
    opts.site = ROOT / 'build' / f'web-{opts.backend}{"" if opts.threads == "1" else "-nothreads"}'
    opts.jobs = max(1, opts.jobs)
    opts.out = Path(opts.out) if opts.out else opts.site / 'webcheck'
    return opts


def first_line(text):
    return text.strip().split('\n')[0]


def check_example(browser, debug_base, base_url, example, opts):
    result = {'example': example, 'errors': [], 'started': False, 'backend_ok': False, 'console': [],
              'pending': -1, 'start_ms': None}
    # Each example gets its own browser context (separate storage, like a fresh
    # profile), so examples checked at the same time don't share the IndexedDB file
    # cache and can't slow or affect each other.
    context = browser.send('Target.createBrowserContext', {'disposeOnDetach': True})['browserContextId']
    target = browser.send('Target.createTarget', {'url': 'about:blank', 'browserContextId': context})['targetId']
    session = open_session(f'ws://{urllib.parse.urlsplit(debug_base).netloc}/devtools/page/{target}')
    lock = threading.Lock()
    inflight = set()
    state = {'navigated': time.monotonic(), 'activity': time.monotonic()}

    def on_event(msg):
        method, params = msg['method'], msg.get('params', {})
        with lock:
            if method == 'Network.requestWillBeSent':
                # a worker's own script (threaded builds start pthread workers) has no
                # loader, and its completion is reported to the worker, not this page
                if not params.get('loaderId'):
                    return
                inflight.add(params['requestId'])
                state['activity'] = time.monotonic()
            elif method in ('Network.loadingFinished', 'Network.loadingFailed'):
                inflight.discard(params['requestId'])
                state['activity'] = time.monotonic()
            elif method == 'Runtime.consoleAPICalled':
                text = ' '.join(str(a['value']) if 'value' in a else a.get('description', '')
                                for a in params['args'])
                result['console'].append(first_line(text))
                if 'libwgrender:' in text and 'backend' in text:
                    result['started'] = True
                    if result['start_ms'] is None:
                        result['start_ms'] = (time.monotonic() - state['navigated']) * 1000
                    state['activity'] = time.monotonic()
                    result['backend_ok'] = result['backend_ok'] or BACKEND_LOG[opts.backend] in text
                # libwgrender logs go to the console as plain messages: fail on error-level
                # ones like tools/smoke.py does ([ERROR], [FATAL])
                if params.get('type') == 'error' or '[panic]' in text or ERROR_LINE.search(text):
                    result['errors'].append(first_line(text))
            elif method == 'Runtime.exceptionThrown':
                d = params['exceptionDetails']
                result['errors'].append(first_line((d.get('exception') or {}).get('description')
                                                   or d.get('text') or 'exception'))
            elif method == 'Log.entryAdded':
                # the browser's own messages: WebGPU validation errors from Dawn, GL driver
                # warnings, network failures. Error-level ones fail the check, except
                # network ones: a missing favicon is a 404 too, and a missing asset
                # already fails through libwgrender's own loading errors
                e = params['entry']
                line = f'[{e["source"]}/{e["level"]}] {first_line(e.get("text") or "")}'
                result['console'].append(line)
                if e['level'] == 'error' and e['source'] != 'network':
                    result['errors'].append(line)

    session.on_event(on_event)
    try:
        for domain in ('Runtime', 'Log', 'Page', 'Network'):
            session.send(f'{domain}.enable')
        deadline = time.monotonic() + opts.settle / 1000
        state['navigated'] = time.monotonic()
        session.send('Page.navigate', {'url': f'{base_url}/?ex={urllib.parse.quote(example)}'})
        # Wait until the example has started, its asset downloads are done, and it has
        # run for a while since (errors from loading show up in that window), or until
        # the deadline, whichever comes first.
        pending = -1
        while time.monotonic() < deadline:
            if not result['started']:
                # don't call into the page before libwgrender reports it's running: calling
                # an exported function before the wasm runtime is initialized aborts the page
                time.sleep(0.1)
                continue
            value = session.send('Runtime.evaluate', {
                'expression': "typeof Module !== 'undefined' && Module._wgri_asset_pending_count"
                              " ? Module._wgri_asset_pending_count() : -1",
                'returnByValue': True})['result'].get('value')
            with lock:
                if value != pending:
                    pending = value
                    state['activity'] = time.monotonic()
                quiet = time.monotonic() - state['activity'] >= opts.quiet / 1000
                done = pending == 0 and not inflight and quiet
            if done:
                break
            time.sleep(0.1)
        result['pending'] = pending
        if pending and pending > 0:  # out of time: have libwgrender say which files, and where they are stuck
            session.try_send('Runtime.evaluate', {
                'expression': "typeof Module !== 'undefined' && Module._wgri_asset_pending_log"
                              " && Module._wgri_asset_pending_log()"})
            time.sleep(0.3)  # the warnings arrive as console events
        shot = session.send('Page.captureScreenshot', {'format': 'png'})
        result['screenshot'] = opts.out / f'{example}.png'
        result['screenshot'].write_bytes(base64.b64decode(shot['data']))
    finally:
        session.close()
        browser.try_send('Target.closeTarget', {'targetId': target})
        browser.try_send('Target.disposeBrowserContext', {'browserContextId': context})
    return result


def problems_of(r, opts):
    problems = list(r['errors'])
    if not r['started']:
        problems.append('never logged its backend (did not start?); last console output:')
        problems += [f'  | {line}' for line in r['console'][-6:]]
    elif r['pending'] is None or r['pending'] < 0:
        problems.append(f'never reported its asset queue in {opts.settle} ms (the runtime was still starting)')
    elif r['pending'] != 0:
        problems.append(f'still loading after {opts.settle} ms ({r["pending"]} asset task(s) pending):')
        stuck = [line for line in r['console'] if 'wgr_asset: pending:' in line]
        problems += [f'  | {line}' for line in (stuck or r['console'][-6:])]
    elif not r['backend_ok']:
        problems.append(f"started on a different backend than '{opts.backend}' (stale build?)")
    return problems


def main():
    opts = parse_args()
    manifest = opts.site / 'examples.json'
    if not manifest.exists():
        sys.exit(f'webcheck: no web build at {opts.site} (build it first: cmake --preset {opts.site.name} '
                 f'&& cmake --build --preset {opts.site.name})')
    built = json.loads(manifest.read_text())
    examples = opts.examples or built
    missing = [e for e in examples if e not in built]
    if missing:
        sys.exit(f'webcheck: not built: {", ".join(missing)}')

    browser_path = find_browser(opts.browser)
    opts.out.mkdir(parents=True, exist_ok=True)
    run = RunProcesses('webcheck')
    run.install_handlers()
    try:
        site_port = free_port()
        run.spawn([PYTHON, ROOT / 'tools' / 'serve.py', site_port, opts.site])
        base_url = f'http://127.0.0.1:{site_port}'
        wait_for(f'{base_url}/examples.json', 'tools/serve.py')
        if opts.display == 'screen' and opts.backend == 'webgpu' and not opts.headed:
            print('webcheck: no Xvfb; WebGPU runs in a window on the real screen')
        debug_base, browser = launch_browser(run, browser_path, opts.display)
        where = 'virtual display (Xvfb)' if opts.display == 'xvfb' else opts.display
        print(f'webcheck: {len(examples)} example(s), backend {opts.backend}, {where}, '
              f'{opts.jobs} at a time, {browser_path}', flush=True)

        # check `jobs` examples at a time; report in order as results complete
        results = [None] * len(examples)
        lock = threading.Lock()
        progress = {'next': 0, 'reported': 0, 'failed': 0}

        def report():
            with lock:
                while progress['reported'] < len(examples) and results[progress['reported']]:
                    r = results[progress['reported']]
                    progress['reported'] += 1
                    problems = problems_of(r, opts)
                    if problems:
                        progress['failed'] += 1
                    start = f' (started after {r["start_ms"] / 1000:.1f} s)' if r.get('start_ms') else ''
                    print(f'  {"FAIL" if problems else "ok  "}  {r["example"]}{start if problems else ""}')
                    for p in problems:
                        print(f'          {p}')
                    if opts.verbose:
                        for line in r['console']:
                            print(f'          | {line}')
                    sys.stdout.flush()

        def worker():
            while True:
                with lock:
                    index = progress['next']
                    if index >= len(examples):
                        return
                    progress['next'] += 1
                try:
                    results[index] = check_example(browser, debug_base, base_url, examples[index], opts)
                except Exception as e:
                    results[index] = {'example': examples[index], 'errors': [f'check failed: {e}'],
                                      'started': True, 'backend_ok': True, 'pending': 0, 'console': []}
                report()

        workers = [threading.Thread(target=worker) for _ in range(min(opts.jobs, len(examples)))]
        for w in workers:
            w.start()
        for w in workers:
            w.join()
        browser.close()
        print(f'screenshots: {opts.out}')
        failed = progress['failed']
        print(f'FAIL: {failed} of {len(examples)} example(s)' if failed else f'PASS: {len(examples)} example(s)')
        return 1 if failed else 0
    finally:
        run.stop()


if __name__ == '__main__':
    try:
        sys.exit(main())
    except RuntimeError as e:
        sys.exit(f'webcheck: {e}')
